### PR TITLE
EIP-3436 - Expanded Clique Block Choice Rule

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Implement the EIP-3436 expanded clique block choice rule.  Specifically, instead of taking the heaviest chain then first observed chain when presented with multiple chain heads the following choices are made in order:

1. Highest Total Difficulty
2. Shortest Chain
3. Most recent in-turn-proposer
4. Lowest Block hash

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java
Comment (RatanRSur) [inline besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java:147]: Started going down a rabbit hole to remove this mutation. I didn't see a way out that isn't too intrusive.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java
Comment (RatanRSur) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java:180]: Could we add an exception here too? Or does it need to be a noop for things to work?

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java
Comment (RatanRSur) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]:
```suggestion
            .thenComparing(header -> distanceFromInTurn(header, cliqueContext)))
```

**Review by RatanRSur — DISMISSED** [review]

**Comment by vmichalik** [comment]
@shemnon is this one on hold or do you need someone else to take a look at it?

**Comment by shemnon** [comment]
I was waiting on some ACD wide consensus on the issue, and the only interest I got from Geth was a comment where they mis-read the spec - https://ethereum-magicians.org/t/eip-3436-expanded-clique-block-choice-rule/5809/5

It's not a change one client should do in isolation, and ACD's agenda is fairly packed.  I expect it to fester until goerli halts again.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java
Comment (shemnon) [reply ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java]: Sure.  Assuming CI doesn't break.

File: besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java
Comment (shemnon) [reply besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java:144]: The only way to be less intrusive is for a more invasive DI change, like going to Dagger.

**Comment by macfarla** [comment]
@shemnon is this one looking for a review or still waiting on ACD?

**Comment by shemnon** [comment]
Still needs that approval I asked for 3 months ago

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java
Comment (macfarla) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:104]: should this be "most recent" or is the PR description wrong?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/Blockchain.java
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/Blockchain.java:238]: s/more preferred/preferred

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java
Comment (macfarla) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]: reading the spec I'm still confused - maybe it's worth putting the formula in for context? 

Clients should choose the block with the largest value for this calculation:

(header_number - validator_index) % validator_count

**Review by macfarla — COMMENTED** [review]
couple of comments. I'd like someone from the mainnet team to review as well, it's a bit out of my area. sonarcloud is complaining but none of that was introduced by this PR

**Review by matkt — APPROVED** [review]
LGTM, just need to resolve @macfarla 's comments

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java
Comment (shemnon) [reply consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]: added javadoc.

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java
Comment (shemnon) [reply consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:104]: this flopped around several times.  current eip says maximum of the formula, so it's an inverted comparator.

**Pull request merged**
